northstar/src/runtime/cgroups.rs

```diff
@@ -32,244 +37,331 @@ use tokio::{
-        .await
-        .expect("Failed to setup event_control");
-    event_control
-        .flush()
-        .await
-        .expect("Failed to setup oom event fd");
-
-    // This task stops when the main loop receiver closes
-    let task = {
-        let stop = stop.clone();
-        task::spawn(async move {
-            debug!("Listening for oom events of {}", container);
-            let mut buffer = [0u8; 1];
-
-            select! {
-                _ = stop.cancelled() => (),
-                _ = tx.closed() => (),
-                _ = event_fd.read(&mut buffer) => {
-                    loop {
-                        match tx.try_send(Event::Oom(container.clone())) {
-                            Ok(_) => break,
-                            Err(TrySendError::Closed(_)) => break,
-                            Err(TrySendError::Full(_)) => time::sleep(time::Duration::from_millis(1)).await,
+#[derive(Debug)]
+struct MemoryMonitor {
+    token: CancellationToken,
+    task: JoinHandle<()>,
+}
+
+impl MemoryMonitor {
+    /// Setup an event fd and oom event listening.
+    async fn new_v1(container: Container, path: &Path, tx: EventTx) -> MemoryMonitor {
+        const OOM_CONTROL: &str = "memory.oom_control";
+        const EVENT_CONTROL: &str = "cgroup.event_control";
+
+        // Configure oom
+        let oom_control = path.join(OOM_CONTROL);
+        let event_control = path.join(EVENT_CONTROL);
+        let token = CancellationToken::new();
+
+        let mut event_fd = EventFd::new(0, false).expect("Failed to create eventfd");
+
+        debug!("Opening oom_control: {}", oom_control.display());
+        let oom_control = fs::OpenOptions::new()
+            .write(true)
+            .open(&oom_control)
+            .await
+            .expect("Failed to open oom_control");
+
+        debug!("Opening event_control: {}", event_control.display());
+        let mut event_control = fs::OpenOptions::new()
+            .write(true)
+            .open(&event_control)
+            .await
+            .expect("Failed to open event_control");
+        event_control
+            .write_all(format!("{} {}", event_fd.as_raw_fd(), oom_control.as_raw_fd()).as_bytes())
+            .await
+            .expect("Failed to setup event_control");
+        event_control
+            .flush()
+            .await
+            .expect("Failed to setup oom event fd");
+
+        // This task stops when the main loop receiver closes
+        let task = {
+            let stop = token.clone();
+            task::spawn(async move {
+                debug!("Listening for oom events of {}", container);
+                let mut buffer = [0u8; 16];
+
+                'outer: loop {
+                    select! {
+                        _ = stop.cancelled() => break 'outer,
+                        _ = tx.closed() => break 'outer,
+                        _ = event_fd.read(&mut buffer) => {
```

Comment:
is something meaningful written to `buffer`? is it just fine to ignore it?

Reply:
Since we're only watching for Modify events the content of buffer is always a modify event with some additional not relevant data. Thinks is safe to ignore.